designs/CentrallyManagingNuGetPackageVersions-VS-APIs.md

```diff
@@ -0,0 +1,419 @@
+* Orchestrate efficient updating and nomination of projects. In particular, when project files and `Directory.Packages.props` files need to change due to a single customer gesture, avoid multiple project nominations and multiple restores due to multiple file changes.
+
+Proposed API.
+
+```cs
+enum PackageReferenceAction
+{
+  /// <summary>
+  /// To catch uninitialized values, but otherwise usused and invalid.
+  /// </summary>
+  Unknown = 0,
+
+  /// <summary>
+  /// The change should add or update PackageReference and or PackageVersion items.
+  /// </summary>
+  AddOrUpdate = 1,
+
+  /// <summary>
+  /// The change should remove matching PackageReference items only, do not remove PackageVersion items.
+  /// </summary>
+  RemovePackageReference = 2,
+
+  /// <summary>
+  /// This change should remove both PackageReference and PackageVersion items.
+  /// </summary>
+  /// <remarks>
+  /// Out of scope for the initial version. It has been defined to avoid needing an API change to add this in the future.
+  /// </remarks>
+  Remove = 3
+}
+
+class PackageReferenceChangeRequest
+{
+  /// <summary>
+  /// The action to apply
+  /// </summary>
+  public PackageReferenceAction Action { get; set;}
+
+  /// <summary>
+  /// Which projects in the solution should apply this PackageReference action
+  /// </summary>
+  public List<string> Projects { get; set; }
+
+  /// <summary>
+  /// The package ID 
+  /// </summary>
+  public string PackageId { get; set; }
+
+  /// <summary>
+  /// The package version
+  /// </summary>
+  /// <remarks>
+  /// * If the project does not use CPVM
+  ///   * AddOrUpdate must set the PackageReference Version metadata to this value
+  ///   * RemovePackage will ignore any value in Version
+  /// * If the project uses CPVM
+  ///   * AddOrUpdate must set the Version attribute in the Directory.Packages.props fie to this version. If the Directory.Packages.props file does not already contain an item for this PackageId, then it should be added.
+  ///   * RemovePackage will ignore any value in Version
+  /// </remarks>
+  public string Version { get; set; }
+
+  /// <summary>
+  /// Additional metadata for this PackageId to set on the PackageReference item.
+  /// </summary>
+  /// <remarks>
+  /// * AddOrUpdate must create or set metadata defined by the keys of this dictionary, with the value defined by the dictionary item's value.
+  /// * AddOrUpdate must remove any metadata already defined in the PackageReference, if there is no key for that metadata in the dictionary.
+  /// * RemovePackage will ignore any value in AdditionalMetadata.
+  /// </remarks>
+  public IDictionary<string, string> AdditionalMetadata { get; set; }
+}
+
+interface IPackageReferenceUpdateService
+{
+  /// <summary>
+  /// Apply changes to the project(s) provided.
+  /// </summary>
+  Task<bool> ApplyChangesAsync(List<PackageReferenceChangeRequest> actions, CancellationToken token);
```

Comment:
If we like to define a list of action to be send to the service, we would need to define:
1. If actions will be applied in order 
2. What are cases of bad requests (if any) and what the expected behavior will be? Cases like request1(project1 package1 version1 addorupdate), request2(project1 package1 version1 removePackageReference) is it bad request or the result is removePackageReference ?